Bun install is failing in vercel (Turbo Repo)

### What version of Bun is running?

1.0.21

### What platform is your computer?

vercel

### What steps can reproduce the bug?

Deploy any project in vercel using bun install (bun.lockb included)

### What is the expected behavior?

The installation should be finished

### What do you see instead?

Error: Cannot find module '/vercel/path0/install.js'

![image](https://github.com/oven-sh/bun/assets/26267698/1d294e5b-67ee-4085-85ed-aa3de2b023ce)


### Additional information

_No response_

Same issue here. I have a Next.js v14.0.4 project in a turbo repo workspace.

```
[07:52:01.423] Running build in Washington, D.C., USA (East) – iad1
[07:52:01.537] Cloning github.com/lmachens/the-hidden-gaming-lair (Branch: main, Commit: c9b4326)
[07:52:01.912] Previous build cache not available
[07:52:03.898] Cloning completed: 2.361s
[07:52:04.628] Running "vercel build"
[07:52:05.089] Vercel CLI 33.0.2
[07:52:05.225] > Detected Turbo. Adjusting default settings...
[07:52:05.794] Running "install" command: `bun install`...
[07:52:05.855] bun install v1.0.21 (837cbd60)
[07:52:08.931] node:internal/modules/cjs/loader:1080
[07:52:08.932]   throw err;
[07:52:08.932]   ^
[07:52:08.932] 
[07:52:08.932] Error: Cannot find module '/vercel/path0/install.js'
[07:52:08.932]     at Module._resolveFilename (node:internal/modules/cjs/loader:1077:15)
[07:52:08.932]     at Module._load (node:internal/modules/cjs/loader:922:27)
[07:52:08.933]     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:86:12)
[07:52:08.933]     at node:internal/main/run_main_module:23:47 {
[07:52:08.933]   code: 'MODULE_NOT_FOUND',
[07:52:08.933]   requireStack: []
[07:52:08.933] }
[07:52:08.933] 
[07:52:08.934] Node.js v18.18.2
[07:52:08.934] 
[07:52:08.934] error: postinstall script from "esbuild" exited with 1
[07:52:08.974] Error: Command "bun install" exited with 1
[07:52:11.429] 
```

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -7688,8 +7688,21 @@ pub const PackageManager = struct {
             }
 
             const trees = this.lockfile.buffers.trees.items;
-            this.tree_install_counts[tree_id] += 1;
-            if (this.tree_install_counts[tree_id] >= trees[tree_id].dependencies.len) {
+            const current_count = this.tree_install_counts[tree_id];
+            const max = trees[tree_id].dependencies.len;
+
+            if (current_count == std.math.maxInt(usize)) {
+                if (comptime Environment.allow_assert)
+                    Output.panic("Installed more packages than expected for tree id: {d}. Expected: {d}", .{ tree_id, max });
+
+                return;
+            }
+
+            const is_not_done = current_count + 1 < max;
+
+            this.tree_install_counts[tree_id] = if (is_not_done) current_count + 1 else std.math.maxInt(usize);
+
+            if (!is_not_done) {
                 this.completed_trees.set(tree_id);
                 this.runAvailableScripts(log_level);
             }
@@ -8771,12 +8784,6 @@ pub const PackageManager = struct {
                 scripts_node.activate();
             }
 
-            if (comptime Environment.allow_assert) {
-                for (this.lockfile.buffers.trees.items) |tree| {
-                    std.debug.assert(installer.tree_install_counts[tree.id] == tree.dependencies.len);
-                }
-            }
-
             if (!installer.options.do.install_packages) return error.InstallFailed;
 
             summary.successfully_installed = installer.successfully_installed;
@@ -9209,6 +9216,8 @@ pub const PackageManager = struct {
             PackageManager.verbose_install or manager.options.do.print_meta_hash_string,
         );
 
+        const should_save_lockfile = did_meta_hash_change or had_any_diffs or needs_new_lockfile or manager.package_json_updates.len > 0;
+
         if (manager.options.global) {
             try manager.setupGlobalDir(&ctx);
         }
@@ -9223,7 +9232,7 @@ pub const PackageManager = struct {
 
         // It's unnecessary work to re-save the lockfile if there are no changes
         if (manager.options.do.save_lockfile and
-            (did_meta_hash_change or manager.lockfile.isEmpty() or manager.options.enable.force_save_lockfile))
+            (should_save_lockfile or manager.lockfile.isEmpty() or manager.options.enable.force_save_lockfile))
         save: {
             if (manager.lockfile.isEmpty()) {
                 if (!manager.options.dry_run) {

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -9248,7 +9248,13 @@ pub const PackageManager = struct {
                     };
                 }
                 if (!manager.options.global) {
-                    if (log_level != .silent) Output.prettyErrorln("No packages! Deleted empty lockfile", .{});
+                    if (log_level != .silent) {
+                        if (manager.to_remove.len > 0) {
+                            Output.prettyErrorln("\npackage.json has no dependencies! Deleted empty lockfile", .{});
+                        } else {
+                            Output.prettyErrorln("No packages! Deleted empty lockfile", .{});
+                        }
+                    }
                 }
 
                 break :save;

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -9186,15 +9186,15 @@ pub const PackageManager = struct {
 
         const needs_clean_lockfile = had_any_diffs or needs_new_lockfile or manager.package_json_updates.len > 0;
         var did_meta_hash_change = needs_clean_lockfile;
-        if (needs_clean_lockfile) {
-            manager.lockfile = try manager.lockfile.cleanWithLogger(
-                manager.package_json_updates,
-                manager.log,
-                manager.options.enable.exact_versions,
-            );
-            if (manager.lockfile.packages.len > 0) {
-                root = manager.lockfile.packages.get(0);
-            }
+
+        // This operation doesn't perform any I/O, so it should be relatively cheap.
+        manager.lockfile = try manager.lockfile.cleanWithLogger(
+            manager.package_json_updates,
+            manager.log,
+            manager.options.enable.exact_versions,
+        );
+        if (manager.lockfile.packages.len > 0) {
+            root = manager.lockfile.packages.get(0);
         }
 
         if (manager.lockfile.packages.len > 0) {
@@ -9208,11 +9208,9 @@ pub const PackageManager = struct {
             manager.lockfile.verifyResolutions(manager.options.local_package_features, manager.options.remote_package_features, log_level);
         }
 
-        if (needs_clean_lockfile or manager.options.enable.force_save_lockfile or manager.options.enable.frozen_lockfile) {
-            did_meta_hash_change = try manager.lockfile.hasMetaHashChanged(
-                PackageManager.verbose_install or manager.options.do.print_meta_hash_string,
-            );
-        }
+        did_meta_hash_change = try manager.lockfile.hasMetaHashChanged(
+            PackageManager.verbose_install or manager.options.do.print_meta_hash_string,
+        );
 
         if (manager.options.global) {
             try manager.setupGlobalDir(&ctx);
@@ -9261,6 +9259,13 @@ pub const PackageManager = struct {
             }
 
             manager.lockfile.saveToDisk(manager.options.lockfile_path);
+
+            if (comptime Environment.allow_assert) {
+                if (manager.lockfile.hasMetaHashChanged(false) catch false) {
+                    Output.panic("Lockfile metahash non-deterministic after saving", .{});
+                }
+            }
+
             if (comptime log_level.showProgress()) {
                 save_node.end();
                 manager.progress.refresh();

--- a/src/install/install.zig
+++ b/src/install/install.zig
@@ -9184,9 +9184,6 @@ pub const PackageManager = struct {
 
         if (manager.log.hasErrors()) Global.crash();
 
-        const needs_clean_lockfile = had_any_diffs or needs_new_lockfile or manager.package_json_updates.len > 0;
-        var did_meta_hash_change = needs_clean_lockfile;
-
         // This operation doesn't perform any I/O, so it should be relatively cheap.
         manager.lockfile = try manager.lockfile.cleanWithLogger(
             manager.package_json_updates,
@@ -9208,7 +9205,7 @@ pub const PackageManager = struct {
             manager.lockfile.verifyResolutions(manager.options.local_package_features, manager.options.remote_package_features, log_level);
         }
 
-        did_meta_hash_change = try manager.lockfile.hasMetaHashChanged(
+        const did_meta_hash_change = try manager.lockfile.hasMetaHashChanged(
             PackageManager.verbose_install or manager.options.do.print_meta_hash_string,
         );
 

--- a/src/install/lockfile.zig
+++ b/src/install/lockfile.zig
@@ -880,6 +880,7 @@ pub fn cleanWithLogger(
     }
     new.trusted_dependencies = old_trusted_dependencies;
     new.scripts = old_scripts;
+    new.meta_hash = old.meta_hash;
 
     return new;
 }

--- a/test/cli/install/bun-add.test.ts
+++ b/test/cli/install/bun-add.test.ts
@@ -1418,9 +1418,8 @@ it("should add dependency without duplication", async () => {
 
   expect(err2).not.toContain("error:");
 
-  // the meta hash did not change, the display name did not change, therefore
-  // the lockfile did not change and should not need to be saved.
-  expect(err2).not.toContain("Saved lockfile");
+  // The meta-hash didn't change, but we do save everytime you do "bun add <package>".
+  expect(err2).toContain("Saved lockfile");
 
   expect(out2.replace(/\s*\[[0-9\.]+m?s\] done\s*$/, "").split(/\r?\n/)).toEqual(["", " installed bar@0.0.2"]);
   expect(await exited2).toBe(0);
@@ -1533,8 +1532,8 @@ it("should add dependency without duplication (GitHub)", async () => {
   const err2 = await new Response(stderr2).text();
   expect(err2).not.toContain("error:");
 
-  // We didn't make any changes to the lockfile, so it should not be saved.
-  expect(err2).not.toContain("Saved lockfile");
+  // The meta-hash didn't change, but we do save everytime you do "bun add <package>".
+  expect(err2).toContain("Saved lockfile");
 
   expect(stdout2).toBeDefined();
   const out2 = await new Response(stdout2).text();

--- a/test/cli/install/bun-add.test.ts
+++ b/test/cli/install/bun-add.test.ts
@@ -983,10 +983,11 @@ it("should install version tagged with `latest` by default", async () => {
   });
   expect(stderr1).toBeDefined();
   const err1 = await new Response(stderr1).text();
+  const out1 = await new Response(stdout1).text();
+
   expect(err1).not.toContain("error:");
   expect(err1).toContain("Saved lockfile");
   expect(stdout1).toBeDefined();
-  const out1 = await new Response(stdout1).text();
   expect(out1.replace(/\s*\[[0-9\.]+m?s\]\s*$/, "").split(/\r?\n/)).toEqual([
     "",
     " installed baz@0.0.3",
@@ -1410,11 +1411,17 @@ it("should add dependency without duplication", async () => {
     env,
   });
   expect(stderr2).toBeDefined();
-  const err2 = await new Response(stderr2).text();
-  expect(err2).not.toContain("error:");
-  expect(err2).toContain("Saved lockfile");
   expect(stdout2).toBeDefined();
+
+  const err2 = await new Response(stderr2).text();
   const out2 = await new Response(stdout2).text();
+
+  expect(err2).not.toContain("error:");
+
+  // the meta hash did not change, the display name did not change, therefore
+  // the lockfile did not change and should not need to be saved.
+  expect(err2).not.toContain("Saved lockfile");
+
   expect(out2.replace(/\s*\[[0-9\.]+m?s\] done\s*$/, "").split(/\r?\n/)).toEqual(["", " installed bar@0.0.2"]);
   expect(await exited2).toBe(0);
   expect(urls.sort()).toBeEmpty();
@@ -1525,7 +1532,10 @@ it("should add dependency without duplication (GitHub)", async () => {
   expect(stderr2).toBeDefined();
   const err2 = await new Response(stderr2).text();
   expect(err2).not.toContain("error:");
-  expect(err2).toContain("Saved lockfile");
+
+  // We didn't make any changes to the lockfile, so it should not be saved.
+  expect(err2).not.toContain("Saved lockfile");
+
   expect(stdout2).toBeDefined();
   const out2 = await new Response(stdout2).text();
   expect(out2.replace(/\s*\[[0-9\.]+m?s\] done\s*$/, "").split(/\r?\n/)).toEqual([

--- a/test/cli/install/bun-remove.test.ts
+++ b/test/cli/install/bun-remove.test.ts
@@ -99,6 +99,8 @@ it("should remove existing package", async () => {
   expect(await removeExited1).toBe(0);
   expect(stdout1).toBeDefined();
   const out1 = await new Response(stdout1).text();
+  const err1 = await new Response(stderr1).text();
+
   expect(out1.replace(/\s*\[[0-9\.]+m?s\]/, "").split(/\r?\n/)).toEqual([
     "",
     ` + pkg2@${pkg2_path}`,
@@ -108,7 +110,6 @@ it("should remove existing package", async () => {
     "",
   ]);
   expect(stderr1).toBeDefined();
-  const err1 = await new Response(stderr1).text();
   expect(err1.replace(/^(.*?) v[^\n]+/, "$1").split(/\r?\n/)).toEqual(["bun remove", " Saved lockfile", ""]);
   expect(await file(join(package_dir, "package.json")).text()).toEqual(
     JSON.stringify(
@@ -139,12 +140,14 @@ it("should remove existing package", async () => {
   expect(await removeExited2).toBe(0);
   expect(stdout2).toBeDefined();
   const out2 = await new Response(stdout2).text();
-  expect(out2.replace(/\s*\[[0-9\.]+m?s\]/, "").split(/\r?\n/)).toEqual([" done", ""]);
-  expect(stderr2).toBeDefined();
   const err2 = await new Response(stderr2).text();
+
+  expect(out2.replace(/\s*\[[0-9\.]+m?s\]/, "").split(/\r?\n/)).toEqual(["", " - pkg2", " 1 package removed", ""]);
+  expect(stderr2).toBeDefined();
   expect(err2.replace(/^(.*?) v[^\n]+/, "$1").split(/\r?\n/)).toEqual([
     "bun remove",
-    "No packages! Deleted empty lockfile",
+    "",
+    "package.json has no dependencies! Deleted empty lockfile",
     "",
   ]);
   expect(await file(join(package_dir, "package.json")).text()).toEqual(

--- a/test/cli/install/overrides.test.ts
+++ b/test/cli/install/overrides.test.ts
@@ -8,9 +8,9 @@ function install(cwd: string, args: string[]) {
   const exec = Bun.spawnSync({
     cmd: [bunExe(), ...args],
     cwd,
-    stdout: "ignore",
-    stdin: "ignore",
-    stderr: "ignore",
+    stdout: "inherit",
+    stdin: "inherit",
+    stderr: "inherit",
     env: bunEnv,
   });
   if (exec.exitCode !== 0) {
@@ -23,9 +23,9 @@ function installExpectFail(cwd: string, args: string[]) {
   const exec = Bun.spawnSync({
     cmd: [bunExe(), ...args],
     cwd,
-    stdout: "ignore",
-    stdin: "ignore",
-    stderr: "ignore",
+    stdout: "inherit",
+    stdin: "inherit",
+    stderr: "inherit",
     env: bunEnv,
   });
   if (exec.exitCode === 0) {
@@ -42,11 +42,12 @@ function versionOf(cwd: string, path: string) {
 
 function ensureLockfileDoesntChangeOnBunI(cwd: string) {
   install(cwd, ["install"]);
-  const lockb_hash = new Bun.CryptoHasher("sha256").update(readFileSync(join(cwd, "bun.lockb"))).digest();
+  const lockb1 = readFileSync(join(cwd, "bun.lockb"));
   install(cwd, ["install", "--frozen-lockfile"]);
   install(cwd, ["install", "--force"]);
-  const lockb_hash2 = new Bun.CryptoHasher("sha256").update(readFileSync(join(cwd, "bun.lockb"))).digest();
-  expect(lockb_hash).toEqual(lockb_hash2);
+  const lockb2 = readFileSync(join(cwd, "bun.lockb"));
+
+  expect(lockb1.toString("hex")).toEqual(lockb2.toString("hex"));
 }
 
 test("overrides affect your own packages", async () => {
@@ -62,7 +63,6 @@ test("overrides affect your own packages", async () => {
   );
   install(tmp, ["install", "lodash"]);
   expect(versionOf(tmp, "node_modules/lodash/package.json")).toBe("4.0.0");
-
   ensureLockfileDoesntChangeOnBunI(tmp);
 });
 

